_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+up into smaller crates that do one thing and do it well**.
+
+Now that we successfully built our crate, what happens if we try to build it again?
+
+```
+$ cargo build
+
+```
+
+Nothing happened at all. Why's that? We can always ask Cargo to give us more
+information through the `--verbose` flag, so let's do that:
+
+```
+$ cargo build --verbose
+       Fresh libc v0.2.10
+       Fresh winapi v0.2.6
+       Fresh winapi-build v0.1.1
+       Fresh kernel32-sys v0.2.2
+       Fresh time v0.1.35
+       Fresh datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Cargo isn't bothering to recompile packages that it knows are "fresh". If you've
+ever had to configure a `Makefile` before, this should be a breath of fresh air
+(no pun intended).
+
+But how does Cargo know that everything is fresh? When Cargo builds a crate, it
+emits a file called `Cargo.lock` that contains the precise versions of all of
+its resolved dependencies:
+
+```
+[root]
+name = "datetime"
+version = "0.1.0"
+dependencies = [
+ "libc 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)",
+ "time 0.1.35 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "kernel32-sys"
+version = "0.2.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "winapi 0.2.6 (registry+https://github.com/rust-lang/crates.io-index)",
+ "winapi-build 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+...
+```
+
+The `Cargo.lock` contains a serialized version of the entire **resolved**
+dependency graph, including precise versions of all of the source code included
+in the build. In the case of a package from crates.io, Cargo stores the name and
+version of the dependency. This is enough information to uniquely identify
+source code from crates.io, because the registry is **append only** (no changes
+to already-published packages are ever allowed).
+
+In addition, the metadata for the registry is stored in a separate git
+repository, and includes checksum for the relevant package. Before Cargo ever
+unpacks a crate it downloads, it first validates the checksum.
+
+Here's what the entry for `kernel32-sys 0.2.2` looks like:
+
+```
+{"name":"kernel32-sys","vers":"0.2.2","deps":[{"name":"winapi","req":"^0.2.5","features":[],"optional":false,"default_features":true,"target":null,"kind":"normal"},{"name":"winapi-build","req":"^0.1.1","features":[],"optional":false,"default_features":true,"target":null,"kind":"build"}],"cksum":"7507624b29483431c0ba2d82aece8ca6cdba9382bff4ddd0f7490560c056098d","features":{},"yanked":false}
+```
+
+### Collaborating
+
+Now for the real test. Let's push our code up to GitHub and develop it on a
+different machine. Ideally, we would like to be able to pick up right where we
+left off, with the exact same source code for all of our dependencies.
+
+To do this, we check in our `Cargo.lock` and clone the repository on our new
```

Comment:
experienced rust user: this is a library crate we aren't supposed to check those in is this bad/what's up with that?